Remove null entry fees.

viroulep opened the pull request:
This removes the possibility for null entry fees, and prevents showing any entry
fees if the value is 0.

Fixes #1468.

jonatanklosko commented on WcaOnRails/app/views/competitions/_competition_info.html.erb:
```diff
@@ -39,7 +39,7 @@
       <% end %>
 
-      <% if competition.base_entry_fee %>
+      <% if competition.base_entry_fee > 0 %>
```
I must point out that this could be `if competition.base_entry_fee.nonzero?`.
That said, whatever you prefer ^^

larspetrus replied:
I'd really prefer a `.has_entry_fee?` function.

 I don't care if it uses `.non_zero` or `> 0` internally.

jonatanklosko commented on WcaOnRails/app/views/registrations/_register_form.html.erb:
```diff
@@ -12,7 +12,8 @@
   <%# If the user has to pay fee here, this information will already be in the payment form %>
-  <% unless @registration.has_to_pay_fee_here? %>
+  <%# If there are no entry fees, no need to diplay this information %>
+  <% unless @registration.has_to_pay_fee_here? || @registration.entry_fee == 0 %>
```
Could be `#zero?` ^^

larspetrus replied:
1. `has_to_pay_fee_here?` needs a name that makes clear what/when/where "here" means.

2. Should maybe the ` || @registration.entry_fee == 0` part be _in_ that function?

viroulep replied:
1: fixed for something clearer.

2: it's actually indirectly in since it uses `outstanding_entry_fees`.
The thing is that we want to show this form "input" if there is any entry fee, but not if the payment form is showing (because the input is already there and we don't need to duplicate the information).

jfly commented:
LGTM!

larspetrus commented on WcaOnRails/db/migrate/20170320222511_change_default_entry_fee_for_competitions.rb:
```diff
@@ -0,0 +1,8 @@
+class ChangeDefaultEntryFeeForCompetitions < ActiveRecord::Migration[5.0]
+  def change
+    Competition.where(base_entry_fee_lowest_denomination: nil).update_all(base_entry_fee_lowest_denomination: 0)
+    change_column_default(:Competitions, :base_entry_fee_lowest_denomination, from: nil, to: 0)
```
Sorry to keep complaining about already reviewed code, but I don't see anything limiting fees to positive numbers. That could be done with a DB constraint, but I think it's better to have a Ruby validation. Unless there is some exotic use case for negative fees?

viroulep commented:
Should have taken care of @larspetrus concerns :)

Bonus commit: I reworked the "You have registered for comp" message, some people mailed the WC team thinking there were "registered" even though the registration was not accepted.

I updated the wording to state "You have submitted your registration for...".

jfly commented on WcaOnRails/app/views/registrations/_register_form.html.erb:
```diff
@@ -12,7 +12,8 @@
 
   <%# If the user has to pay fee here, this information will already be in the payment form %>
-  <% unless @registration.has_to_pay_fee_here? %>
+  <%# If there are no entry fees, no need to diplay this information %>
```
nit: diplay -> display

jfly commented on WcaOnRails/app/views/registrations/_register_form.html.erb:
```diff
@@ -12,7 +12,8 @@
   <%# If the user has to pay fee here, this information will already be in the payment form %>
-  <% unless @registration.has_to_pay_fee_here? %>
+  <%# If there are no entry fees, no need to diplay this information %>
+  <% unless @registration.show_payment_form? || @registration.entry_fee == 0 %>
```
@larspetrus may have already said some stuff that supercedes this, but it looks like you could use the `has_entry_fee?` method here instead of the `== 0` comparison. I'd probably rewrite it as:

```
if !registration.show_payment_form? && @registration.has_entry_fee?
```

viroulep replied:
I'm dumb, I created it (almost) just for that...

viroulep commented:
> Only remaining thing is the here part of has_to_pay_fee_here?. Could it just be has_to_pay_fee? ?
Not really: the competitor could have to pay fee either "here" (=on the WCA website) or on some other platform.

We could change the name to be more explicit that it's actually `has_to_pay_fee_with_stripe`.

larspetrus commented:
Oh, _that's_ what "here" means!! My best guess was that it meant "on this form".

I don't think `...with_stripe` is ideal because you can pay with Stripe on many other sites. And we may add non Stripe payment some day.

I'd call it something like `is_paid_through_wca?`.  

Because 
- `registration.has_to_pay...` reads wrong. A user pays, but a registration does not.
- "pay fee" is a bit redundant. 
- `...through_wca` is one of several possible ways to clearly say "using this site".

viroulep commented:
>I'd call it something like `is_paid_through_wca?`.

I almost agree with this name!
"is_paid" would sound like the registration is already paid, but actually the method indicates if it needs to be paid.
So `to_be_paid_through_wca?` sounds more accurate to me.

larspetrus commented:
Yeah, you're right. That's a better name.

larspetrus commented:
I meant that as a LGTM, since I had earlier said that was the only problem I had.

Maybe I need to be more explicit?

viroulep commented:
> I meant that as a LGTM, since I had earlier said that was the only problem I had.
Maybe I need to be more explicit?

Nop that was fine, I was just kind of busy and couldn't find the time to properly deploy it :)
